File: internal/audio/synthesizer_test.go
package audio

import (
	"testing"
)

func TestSynthesizerPolyphony(t *testing.T) {
	synth := NewSynthesizer()

	if synth.Polyphony() != 8 {
		t.Errorf("Expected default polyphony to be 8, got %d", synth.Polyphony())
	}

	synth.SetPolyphony(16)
	if synth.Polyphony() != 16 {
		t.Errorf("Expected polyphony to be 16 after setting, got %d", synth.Polyphony())
	}
}

func TestSynthesizerNoteOnOff(t *testing.T) {
	synth := NewSynthesizer()
	synth.SetPolyphony(2)

	synth.NoteOn(1, 60, 100) // Middle C
	synth.NoteOn(1, 64, 100) // E
	synth.NoteOn(1, 67, 100) // G (should be ignored due to polyphony)

	synth.mu.Lock()
	activeCount := len(synth.activeNotes)
	synth.mu.Unlock()

	if activeCount != 2 {
		t.Errorf("Expected 2 active notes, got %d", activeCount)
	}

	synth.NoteOff(1, 60)

	synth.mu.Lock()
	activeCount = len(synth.activeNotes)
	synth.mu.Unlock()

	if activeCount != 1 {
		t.Errorf("Expected 1 active note after NoteOff, got %d", activeCount)
	}
}

func TestSynthesizerProcessBlock(t *testing.T) {
	synth := NewSynthesizer()
	buffer := make([]float32, 256)

	// Process without any notes, buffer should be empty
	synth.Process(buffer)
	for i, sample := range buffer {
		if sample != 0.0 {
			t.Fatalf("Expected buffer to be empty, but got %f at index %d", sample, i)
		}
	}

	// Process with an active note
	synth.NoteOn(1, 60, 100)
	synth.Process(buffer)

	nonZeroFound := false
	for _, sample := range buffer {
		if sample != 0.0 {
			nonZeroFound = true
			break
		}
	}

	if !nonZeroFound {
		t.Fatalf("Expected buffer to contain non-zero audio data after Process with active note")
	}
}

File: rust/src/audio/synthesizer_test.rs
use super::synthesizer::OmniSynthesizer;
use super::audio_graph::AudioNode;

#[test]
fn test_synthesizer_polyphony() {
    let synth = OmniSynthesizer::new();
    assert_eq!(synth.polyphony(), 8);

    synth.set_polyphony(16);
    assert_eq!(synth.polyphony(), 16);
}

#[test]
fn test_synthesizer_note_on_off() {
    let synth = OmniSynthesizer::new();
    synth.set_polyphony(2);

    synth.note_on(1, 60, 100);
    synth.note_on(1, 64, 100);
    synth.note_on(1, 67, 100); // Should be ignored

    assert_eq!(synth.active_notes_count(), 2);

    synth.note_off(1, 60);

    assert_eq!(synth.active_notes_count(), 1);
}

#[test]
fn test_synthesizer_process_block() {
    let mut synth = OmniSynthesizer::new();
    let mut buffer = vec![0.0; 256];

    synth.process(&mut buffer);
    for sample in buffer.iter() {
        assert_eq!(*sample, 0.0);
    }

    synth.note_on(1, 60, 100);
    synth.process(&mut buffer);

    let mut non_zero_found = false;
    for sample in buffer.iter() {
        if *sample != 0.0 {
            non_zero_found = true;
            break;
        }
    }

    assert!(non_zero_found, "Expected buffer to contain non-zero audio data");
}
